ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/BackingStore.hs

```diff
@@ -0,0 +1,354 @@
+      -- disk.
+      --
+      -- FIXME: can we satisfy this invariant if we read keys from disk but all
+      -- of them were deleted in the changelog?
+    , rqCount :: !Int
+    }
+    deriving stock (Show, Eq)
+
+deriving via OnlyCheckWhnfNamed "BackingStoreValueHandle" (BackingStoreValueHandle m keys values)
+  instance NoThunks (BackingStoreValueHandle m keys values)
+
+-- | A combination of 'bsValueHandle' and 'bsvhRead'
+bsRead ::
+     IOLike m
+  => BackingStore m keys values diff
+  -> keys
+  -> m (WithOrigin SlotNo, values)
+bsRead store keys = withBsValueHandle store $ \slot vh -> do
+    values <- bsvhRead vh keys
+    pure (slot, values)
+
+-- | A 'IOLike.bracket'ed 'bsValueHandle'
+withBsValueHandle ::
+     IOLike m
+  => BackingStore m keys values diff
+  -> (WithOrigin SlotNo -> BackingStoreValueHandle m keys values -> m a)
+  -> m a
+withBsValueHandle store kont =
+    IOLike.bracket
+      (bsValueHandle store)
+      (bsvhClose . snd)
+      (uncurry kont)
+
+{-------------------------------------------------------------------------------
+  An in-memory backing store
+-------------------------------------------------------------------------------}
+
+data TVarBackingStoreContents m values =
+    TVarBackingStoreContentsClosed
+  | TVarBackingStoreContents
+      !(WithOrigin SlotNo)
+      !values
+  deriving (Generic, NoThunks)
+
+data TVarBackingStoreExn =
+    TVarBackingStoreClosedExn
+  | TVarBackingStoreValueHandleClosedExn
+  | TVarBackingStoreDirectoryExists
+  | TVarBackingStoreNonMonotonicSeq !(WithOrigin SlotNo) !(WithOrigin SlotNo)
+  | TVarBackingStoreDeserialiseExn CBOR.DeserialiseFailure
+  | TVarIncompleteDeserialiseExn
+  deriving anyclass (Exception)
+  deriving stock    (Show)
+
+newtype StoreDirIsIncompatible = StoreDirIsIncompatible FS.FsErrorPath
+  deriving anyclass (Exception)
+
+instance Show StoreDirIsIncompatible where
+  show (StoreDirIsIncompatible p) =
+       "In-Memory database not found in the database directory: "
+    <> show p
+    <> ".\nPre-UTxO-HD and LMDB implementations are incompatible with the In-Memory \
+       \ implementation. Please delete your ledger database directory."
+
+-- | Use a 'TVar' as a trivial backing store
+newTVarBackingStoreInitialiser ::
+     (IOLike m, NoThunks values)
+  => (keys -> values -> values)
+  -> (RangeQuery keys -> values -> values)
+  -> (values -> diff -> values)
+  -> (values -> CBOR.Encoding)
+  -> (forall s. CBOR.Decoder s values)
+  -> BackingStoreInitialiser m keys values diff
+newTVarBackingStoreInitialiser lookup_ rangeRead_ forwardValues_ enc dec =
+  BackingStoreInitialiser $ \(FS.SomeHasFS fs0) initialization -> do
+    ref <- do
+      (slot, values) <- case initialization of
+        InitFromCopy (BackingStorePath path) -> do
+          tvarFileExists <- FS.doesFileExist fs0 (extendPath path)
+          unless tvarFileExists $
+            IOLike.throwIO . StoreDirIsIncompatible $ FS.mkFsErrorPath fs0 path
```

Comment:
Suggestion: Given the  number of uses of qualified calls of various functions from `IOLike` and `FS`, use explicit imports instead to unclutter the code and ease the task of the future reader (which could be yourself :) )

Reply:
Done!